File: app/src/main/java/com/ayan/ritual/render/Palette.kt
package com.ayan.ritual.render

/**
 * Flat, warm, high-contrast. No gradients, no glow, no shadow.
 *
 * A kept day is solid ink on a colour block; a missed one is a wash of that
 * same ink. Nothing emits, so a dense year stays legible where a glowing grid
 * smears into blobs.
 */
object Palette {
    val CREAM = 0xFFE7E3D4.toInt()
    val PAPER = 0xFFF4F2EA.toInt()
    val WHITE = 0xFFFFFFFF.toInt()
    val INK = 0xFF12120F.toInt()

    /** Ink at the weights the grid uses. */
    val INK_MISSED = 0x3312120F
    val INK_FUTURE = 0x1712120F
    val INK_RULE = 0x2E12120F
    val INK_SOFT = 0x8F12120F
}

/**
 * A ritual's colour. [block] is the card, [onBlock] everything drawn on it.
 * [inverted] is the same colour used as ink once the day is marked and the
 * card flips to black.
 */
data class Accent(
    val name: String,
    val block: Int,
    val onBlock: Int = Palette.INK
)

val ACCENTS: List<Accent> = listOf(
    Accent("Lime", 0xFFC9F73F.toInt()),
    Accent("Red", 0xFFE5331C.toInt(), Palette.PAPER),
    Accent("Sage", 0xFF9DB8A4.toInt()),
    Accent("Camel", 0xFFBC9F76.toInt()),
    Accent("Butter", 0xFFEDE55C.toInt()),
    Accent("Orange", 0xFFF26A1B.toInt())
)

fun accentAt(index: Int): Accent = ACCENTS[((index % ACCENTS.size) + ACCENTS.size) % ACCENTS.size]

/** Applies an alpha (0..255) to a packed ARGB colour, preserving its RGB. */
fun Int.withAlpha(alpha: Int): Int =
    (this and 0x00FFFFFF) or ((alpha.coerceIn(0, 255)) shl 24)

/** Linear blend between two packed ARGB colours. */
fun blend(a: Int, b: Int, t: Float): Int {
    val f = t.coerceIn(0f, 1f)
    fun ch(shift: Int): Int {
        val av = (a shr shift) and 0xFF
        val bv = (b shr shift) and 0xFF
        return (av + (bv - av) * f).toInt().coerceIn(0, 255)
    }
    return (ch(24) shl 24) or (ch(16) shl 16) or (ch(8) shl 8) or ch(0)
}

File: ios/Shared/Palette.swift
import UIKit

/**
 Flat, warm, high-contrast. No gradients, no glow, no shadow.

 A kept day is solid ink on a colour block; a missed one is a wash of that same
 ink. Nothing emits, so a dense year stays legible where a glowing grid smears
 into blobs.
 */
enum Palette {
    static let cream: ARGB = 0xFFE7E3D4
    static let paper: ARGB = 0xFFF4F2EA
    static let white: ARGB = 0xFFFFFFFF
    static let ink: ARGB   = 0xFF12120F

    /// Ink at the weights the grid uses.
    static let inkMissed: ARGB = 0x3312120F
    static let inkFuture: ARGB = 0x1712120F
    static let inkRule: ARGB   = 0x2E12120F
    static let inkSoft: ARGB   = 0x8F12120F
}

/// A packed 32-bit colour, alpha first — the same literals the Kotlin uses.
typealias ARGB = UInt32

extension ARGB {

    var uiColor: UIColor {
        UIColor(
            red: CGFloat((self >> 16) & 0xFF) / 255.0,
            green: CGFloat((self >> 8) & 0xFF) / 255.0,
            blue: CGFloat(self & 0xFF) / 255.0,
            alpha: CGFloat((self >> 24) & 0xFF) / 255.0
        )
    }

    var cgColor: CGColor { uiColor.cgColor }

    /// The same colour at a new alpha (0...255), preserving its RGB.
    func withAlpha(_ alpha: Int) -> ARGB {
        let a = UInt32(min(max(alpha, 0), 255))
        return (self & 0x00FFFFFF) | (a << 24)
    }
}

/**
 A ritual's colour. `block` is the card, `onBlock` everything drawn on it.
 Once today is marked the card flips to ink and `block` becomes the grid's ink.
 */
struct Accent: Equatable {
    let name: String
    let block: ARGB
    let onBlock: ARGB

    init(_ name: String, _ block: ARGB, _ onBlock: ARGB = Palette.ink) {
        self.name = name
        self.block = block
        self.onBlock = onBlock
    }
}

let ACCENTS: [Accent] = [
    Accent("Lime", 0xFFC9F73F),
    Accent("Red", 0xFFE5331C, Palette.paper),
    Accent("Sage", 0xFF9DB8A4),
    Accent("Camel", 0xFFBC9F76),
    Accent("Butter", 0xFFEDE55C),
    Accent("Orange", 0xFFF26A1B)
]

func accentAt(_ index: Int) -> Accent {
    ACCENTS[((index % ACCENTS.count) + ACCENTS.count) % ACCENTS.count]
}
